plugin/sampling/strategystore/adaptive/processor.go

```diff
@@ -0,0 +1,513 @@
+// qps. This function assumes that the most recent qps is at the head of the slice.
+func (p *processor) calculateWeightedQPS(allQPS []float64) float64 {
+	if len(allQPS) == 0 {
+		return 0
+	}
+	weights := p.weightsCache.getWeights(len(allQPS))
+	var qps float64
+	for i := 0; i < len(allQPS); i++ {
+		qps += allQPS[i] * weights[i]
+	}
+	return qps
+}
+
+func (p *processor) prependServiceCache() {
+	p.serviceCache = append([]samplingCache{make(samplingCache)}, p.serviceCache...)
+	if len(p.serviceCache) > serviceCacheSize {
+		p.serviceCache = p.serviceCache[0:serviceCacheSize]
+	}
+}
+
+func (p *processor) calculateProbabilitiesAndQPS() (model.ServiceOperationProbabilities, model.ServiceOperationQPS) {
+	p.prependServiceCache()
+	retProbabilities := make(model.ServiceOperationProbabilities)
+	retQPS := make(model.ServiceOperationQPS)
+	svcOpQPS := p.generateOperationQPS()
+	totalOperations := int64(0)
+	for svc, opQPS := range svcOpQPS {
+		if _, ok := retProbabilities[svc]; !ok {
+			retProbabilities[svc] = make(map[string]float64)
+		}
+		if _, ok := retQPS[svc]; !ok {
+			retQPS[svc] = make(map[string]float64)
+		}
+		for op, qps := range opQPS {
+			totalOperations++
+			avgQPS := p.calculateWeightedQPS(qps)
+			retQPS[svc][op] = avgQPS
+			retProbabilities[svc][op] = p.calculateProbability(svc, op, avgQPS)
+		}
+	}
+	p.operationsCalculatedGauge.Update(totalOperations)
+	return retProbabilities, retQPS
+}
+
+func (p *processor) calculateProbability(service, operation string, qps float64) float64 {
+	oldProbability := p.DefaultSamplingProbability
+	// TODO: is this loop overly expensive?
+	p.RLock()
+	if opProbabilities, ok := p.probabilities[service]; ok {
+		if probability, ok := opProbabilities[operation]; ok {
+			oldProbability = probability
+		}
+	}
+	latestThroughput := p.throughputs[0].throughput
+	p.RUnlock()
+
+	usingAdaptiveSampling := p.usingAdaptiveSampling(oldProbability, service, operation, latestThroughput)
+	p.serviceCache[0].Set(service, operation, &samplingCacheEntry{
+		probability:    oldProbability,
+		usingAdapative: usingAdaptiveSampling,
+	})
+
+	// Short circuit if the qps is close enough to targetQPS or if the service isn't using
+	// adaptive sampling.
+	if math.Abs(qps-p.TargetQPS) < p.QPSEquivalenceThreshold || !usingAdaptiveSampling {
+		return oldProbability
+	}
+	var newProbability float64
+	if floatEquals(qps, 0) {
+		// Edge case, we double the sampling probability if the QPS is 0 so that we force the service
+		// to at least sample one span probabilistically
+		newProbability = oldProbability * 2.0
+	} else {
+		newProbability = p.probabilityCalculator.Calculate(p.TargetQPS, qps, oldProbability)
+	}
+	return math.Min(maxSamplingProbability, math.Max(p.MinSamplingProbability, newProbability))
+}
+
+func combineProbabilities(p1 map[string]struct{}, p2 map[string]struct{}) map[string]struct{} {
+	for probability := range p2 {
+		p1[probability] = struct{}{}
+	}
+	return p1
```

Comment:
doesn't seem to combine anything, only ensure keys(p1) are a superset of keys(p2)

also, calling a key "probability" looks weird, how is string a probability?